client/components/credit-card/stored-card.tsx

```diff
@@ -29,7 +31,10 @@ export const getCreditCardSummary = ( translate, type, digits ) => {
 		visa: translate( 'VISA' ),
 	};
 
-	const displayType = supportedTypes[ type && type.toLowerCase() ] || type;
+	const displayType =
+		( supportedTypes as typeof supportedTypes & { [k: string]: undefined } )[
```

Comment:
I don't fully grok this (blame my weak TS knowledge), but it seems ok.

Reply:
I'm going to guess what happened here…

We constrain `type` as a `string` instead of as a key to the supported types. Because of that when we select the `displayType` we get an error saying that `string` does not satisfy the constraints on `supportedTypes`.

So we take the error message and find a way to make it work. The fundamental problem is that `supportedTypes` may or may not contain `type` and we are accepting `type` knowing it may not be in `supportedTypes`.

There are solutions to this:

On the one hand we can type `type` as `keyof supportedTypes` and require callees to constrain it so. That would be fair and appropriate if everything happens within Calypso. If we're getting values from the backend though and they may not be one of these values we're in a pickle because _somewhere_ we're getting in a `string` (though the argument can be made that we should parse that API response and produce a `CreditCardTypeName`)

Our other option is to lean more heavily on TypeScript's inference and remove what looks like a workaround (the type is the type of this other thing _and_ possibly some other keys that don't exist)

```ts
const typeId = type && type.toLowerCase();
const displayType = typeId && supportedTypes.hasOwnProperty( typeId )
	? supportedTypes[ typeId ]
	: type;
```

Probably worth coming back to this in the code

Reply:
I've pushed 07ea52f1c3eea7f97a1e492d42dfc853e76fd071 which inlcudes the `switch` version.

Unfortunately, that introduces type issues around `translate` which may not return a string. This seems acceptable, it's a problem across Calypso. https://github.com/Automattic/wp-calypso/pull/37040 has some related work and conversation.

Reply:
I think that's much more clear. Glad I said something! 😁